### PR TITLE
Modified serilization to support units saved as IComparable

### PR DESCRIPTION
**Pull request by eriove**
I had a class that I needed to serialize that had a unit as a IComparable. This pull requests makes this possible. I've tried testing other alternatives to not break anything. Let me know if you can think of any other use cases that should be tested.

### CODE REVIEW

File: UnitsNet.Serialization.JsonNet/UnitsNetJsonConverter.cs [hunk 1]
```diff
@@ -184,7 +225,12 @@ public override bool CanConvert(Type objectType)
                 return CanConvertNullable(objectType);
             }
 
-            return objectType.Namespace != null && objectType.Namespace.Equals("UnitsNet");
+            return objectType.Namespace != null && 
+                (objectType.Namespace.Equals("UnitsNet") || 
+                objectType == typeof(ValueUnit) || 
+                // All unit types implement IComparable
+                objectType==typeof(IComparable) ||
+                objectType.FullName.StartsWith("System." + nameof(IComparable)));
```
Comment (eriove) [inline UnitsNet.Serialization.JsonNet/UnitsNetJsonConverter.cs]: This was very much trial and error and a later refactoring made this unnecessary.

**Comment by eriove** [comment]
`TypeNameHandling.Auto` is needed for Json.Net to serialize the type information. By doing that no extra code is needed to handle deserialization of properties with interfaces. This is very convenient in combination with a plugin framework where each plugin requires it's own setting (implementing the same basic interface).

I've extracted the serializer code. Let me know if it is clear enough.

After refactoring my tests I realized that I didn't need `TypeNameHandling.All`.

Added the requested test cases, trying all different combinations with 3 `IComparable`.

Added the non UnitsNet `IComparable`.

File: UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs [hunk 2]
```diff
@@ -226,12 +227,185 @@ public void UnitEnumChangedAfterSerialization_ExpectUnitCorrectlyDeserialized()
...
+                            {
+                                result.Add(new object[] { objects[i], objects[j], objects[k]});
+                            }
+                        }
+                    }
+                    return result.ToArray();
+                }
+            }
+
+            private static JsonSerializerSettings CreateJsonSerializerSettigns()
```
Comment (anjdreas) [inline UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs]: Typo

File: UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs [hunk 3]
```diff
@@ -226,12 +227,185 @@ public void UnitEnumChangedAfterSerialization_ExpectUnitCorrectlyDeserialized()
...
+
+                string json = JsonConvert.SerializeObject(testObjWithValueAndUnit, jsonSerializerSettings);
+                TestObjWithValueAndUnit deserializedTestObject = JsonConvert.DeserializeObject<TestObjWithValueAndUnit>(json, jsonSerializerSettings);
+
+                Assert.That(deserializedTestObject.Value.GetType(), Is.EqualTo(typeof(double)));
+                Assert.That(deserializedTestObject.Value, Is.EqualTo(5.0));
+                Assert.That(deserializedTestObject.Unit, Is.EqualTo("Test"));
+            }
+
+            [Test, TestCaseSource(nameof(TestObjectsForThreeObjectsInIComparableWithDifferentValues_ExpectAllCorrectlyDeserialized))]
```
Comment (anjdreas) [inline UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs]: This part was slightly confusing to me, as I haven't used TestCaseSource before. If there is only one case, I think I would prefer a normal method that give me that `TestObjWithThreeIComparable` object.

File: UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs [hunk 4]
```diff
@@ -226,12 +227,185 @@ public void UnitEnumChangedAfterSerialization_ExpectUnitCorrectlyDeserialized()
...
+                    Value2 = comparable2,
+                    Value3 = comparable3,
+                };
+                JsonSerializerSettings jsonSerializerSettings = CreateJsonSerializerSettigns();
+
+                string json = JsonConvert.SerializeObject(testObjWithIComparable, jsonSerializerSettings);
+                var deserializedTestObject = JsonConvert.DeserializeObject<TestObjWithThreeIComparable>(json, jsonSerializerSettings);
+
+                Assert.That(deserializedTestObject.Value1.GetType(), Is.EqualTo(comparable1.GetType()));
+                Assert.That(((deserializedTestObject.Value1)), Is.EqualTo(comparable1));
```
Comment (anjdreas) [inline UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs]: Redundant parantheses.

File: UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs [hunk 5]
```diff
@@ -226,12 +227,185 @@ public void UnitEnumChangedAfterSerialization_ExpectUnitCorrectlyDeserialized()
...
+                Assert.That(deserializedTestObject.Value2.GetType(), Is.EqualTo(comparable2.GetType()));
+                Assert.That(((deserializedTestObject.Value2)), Is.EqualTo(comparable2));
+                Assert.That(deserializedTestObject.Value3.GetType(), Is.EqualTo(comparable3.GetType()));
+                Assert.That(((deserializedTestObject.Value3)), Is.EqualTo(comparable3));
+            }
+
+            public static object[] TestObjectsForThreeObjectsInIComparableWithDifferentValues_ExpectAllCorrectlyDeserialized
```
Comment (anjdreas) [inline UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs]: Can be private.

File: UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs [hunk 6]
```diff
@@ -226,12 +227,185 @@ public void UnitEnumChangedAfterSerialization_ExpectUnitCorrectlyDeserialized()
...
+            }
         }
 
         internal class TestObj
         {
             public Frequency? NullableFrequency { get; set; }
             public Frequency NonNullableFrequency { get; set; }
         }
+
+        internal class TestObjWithValueAndUnit : IComparable
```
Comment (anjdreas) [inline UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs]: All these types can be private, can't they?

File: UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs [hunk 7]
```diff
@@ -226,12 +227,185 @@ public void UnitEnumChangedAfterSerialization_ExpectUnitCorrectlyDeserialized()
...
+            [Test]
+            public void UnitInIComparable_ExpectUnitCorrectlyDeserialized()
+            {
+                TestObjWithIComparable testObjWithIComparable = new TestObjWithIComparable()
+                {
+                    Value = Power.FromWatts(10)
+                };
+                JsonSerializerSettings jsonSerializerSettings = CreateJsonSerializerSettigns();
+
+                string json = JsonConvert.SerializeObject(testObjWithIComparable,jsonSerializerSettings);
```
Comment (anjdreas) [inline UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs]: Add space between params.

File: UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs (hunk 3 above)
Comment (eriove) [inline UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs]: This will actually be 9 tests (testing all combinations of objects). Might be a little too much. I started with 3 very similar tests and did this to reduce code duplication and then it was easy to add the loop with all combinations.

File: UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs (hunk 6 above)
Comment (eriove) [inline UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs]: That's true. Followed the existing class, will change all of them.

**Comment by eriove** [comment]
Fixed the typos and formatting. Kept the TestCaseSource. Let's discuss further.

**Comment by anjdreas** [comment]
Will try to get to this tomorrow.

File: UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs (hunk 3 above)
Comment (anjdreas) [inline UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs]: Ah, ok. 27 tests seem a bit overboard, yes. Tests are nice, but I want each test to bring value and from the way I understand the implementation design and the test, I don't see how testing multiple variations of `TestObjWithThreeIComparable` will help us avoid bugs here. I think I'd rather have one or possibly two test cases with different combinations, but that should be quite enough, I think.

Their docs helped me understand what goes on: http://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_TypeNameHandling.htm

Basically they add a `$type` prop that describes the actual object type when it was serialized, then the deserializer tries to use the exact same type by its namespace + name.

Comment (anjdreas) [inline UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs]: Oops, wrong link: http://www.newtonsoft.com/json/help/html/serializetypenamehandling.htm

Comment (anjdreas) [inline UnitsNet.Serialization.JsonNet.Tests/UnitsNetJsonConverterTests.cs]: Oops, wrong link: http://www.newtonsoft.com/json/help/html/serializetypenamehandling.htm

**Comment by eriove** [comment]
I removed the `TestCaseSource` and just kept one test that tests with `TestObjWithThreeIComparable`.

**Comment by anjdreas** [comment]
Nuget 1.0.0-alpha5 is now out.
